submission-panels.md

```diff
@@ -0,0 +1,69 @@
+# Submission-panels Endpoints
+[Back to the list of all defined endpoints](endpoints.md)
+
+## Main Endpoint
+**/api/configuration/submission-definitions**   
+
+Provide access to the configured submission definitions. It returns the list of existent submission-definitions.
+
+Example: to be provided
+
+## Single Submission-panels 
+**/api/configuration/submission-definitions/<:definition-name>**
+
+Provide detailed information about a specific submission-definition. The JSON response document is as follow
+```json
+{
+  "name": "traditional",
+  "panels": [
+  {
+  	header: {"en": "First page"},
+  	mandatory: true,
+  	type: "inputform",
+  	scope: null
+  },
+  {
+  	header: {"en": "Second page"},
+  	mandatory: true,
+  	type: "inputform",
+  	scope: null
+  },
+  {
+  	header: {"en": "Files and access condition"},
+  	mandatory: true,
+  	type: "uploadWithEmbargo",
```

Comment:
What is the implication of this "type" field? Does this mean that (any future) UI needs to know how to render an "uploadWithEmbargo" panel? Or is it possible to retrieve the "panel definition" of that type though another REST endpoint?

An alternative would be to abstract away the difference between "input-forms" and "submission-definitions"? For example: 
* `/api/submission` -> List of collection IDs in which you are authorized to submit 
* `/api/submission/<coll-id>` -> List of panels (steps) for a submission in this collection 
* `/api/submission/<coll-id>/<panel-number>` -> JSON representation of all the fields with their description present on the current panel.

For entering metadata this would be
 ```
 header: {"en": "Describe part 1"},
 fields: [
   {
     label: {"en":"Authors"},
     repeatable: false,
     mandatory: null,
     description: {"en":"Enter the names of the authors of this item."},
     input: {
       type: "name"
     }
     scope: null, 
     visibility: {
       main: null,
       other: null
     },
     typeBind: []
   },
   {
   label: {"en":"Title"}
     repeatable: false,
     mandatory: {"en":"You must enter a main title for this item."},
     description: "Enter the main title of the item.",
     input: {
       type: "onebox",
       regex: null,
       closedVocabulary: false                  
     }
     scope: null, 
     visibility: {
       main: null,
       other: null
     },
     typeBind: []
   },
   ...
 ]
 ```
For a file upload step this would be
 ```
 header: {"en": "Upload file"},
 fields: [
   {
     label: {"en":"Select or drop file"},
     repeatable: true,
     mandatory: null,
     description: {"en":"Select or drag and drop a file."},
     input: {
       type: "file"
     }
     scope: null, 
     visibility: {
       main: null,
       other: null
     },
     typeBind: []
   },
   {
   label: {"en":"Embargo date"}
     repeatable: false,
     mandatory: null,
     description: "If this file has an embargo, please select the end date here",
     input: {
       type: "date",
     }
     scope: null, 
     visibility: {
       main: null,
       other: null
     },
     typeBind: []
   },
   ...
 ]
 ```
For a license step this would be
 ```
 header: {"en": "Repository license"},
 fields: [
   {
     label: {"en":"Repository license"},
     repeatable: true,
     mandatory: null,
     description: {"en":"Repository license comes here"},
     input: null,
     scope: null, 
     visibility: {
       main: null,
       other: null
     },
     typeBind: []
   },
   {
     label: {"en":"Agreement"},
     repeatable: true,
     mandatory: null,
     description: {"en":"Check this box if you accept the repository license"},
     input: {
       type: "checkbox"
     },
     scope: null, 
     visibility: {
       main: null,
       other: null
     },
     typeBind: []
   }
   ...
 ]
 ```

Reply:
@tomdesair asks a good question here.  I agree that we don't need to exactly replicate the model of "input-forms" (i.e. input-forms.xml) and "submission-panels" (i.e. steps in item-submission.xml), as doing so actually could hinder our ability to change the backend storage of these configs in the future.  

As "input-forms" are really only used to model the structure of "Describe" pages (panels), it seems like we could potentially drop that concept/endpoint altogether for a model like @tomdesair proposes above ⬆️

Reply:
@tdonohue @tomdesair I agree that at the end could be nice to end with a single endpoint to expose "all the submission configuration" but this could be done eventually in an iterative way, so that we can evaluate at each stage if we are able to achieve the next goal within the dspace 7 timeframe. Right now we have these two different, separated but interconnected configuration files so I don't think that merge them is something that must be absolutely done in DSpace 7. Moreover it could be easier to have a basic panel definition that eventually refer (HAL link) a specific configuration (input-form, embargo definition, etc.) instead than a hierarchy of objects that share a common base class (AbstractPanelConfiguration or some like that). Indeed, as we are using the spring data repository approach it will be easier to have separate endpoint for each entity.

In addition keep in mind that the StepConfig (aka the panel definition in this proposal) is one of the main extension point for the submission system. Institution are allowed to write their own implementation that could require very different configuration beyond our ability to describe a form.

> What is the implication of this "type" field? Does this mean that (any future) UI needs to know how to render an "uploadWithEmbargo" panel? Or is it possible to retrieve the "panel definition" of that type though another REST endpoint?

yes the type property need to be understood by the REST client and treated in an appropriate way. The panels have very complex and different behaviors that cannot be (in a simple way) generated automatically from a configuration file. Instead to write complex generic implementation I really think that it will be easier to implement the 5-6 very specialized panels that we need and allow institution to customize the system introducing their own panel with a custom implementation not bound to a predefined, limited configuration schema. If the panel type will require further configuration (as in the case of the inputform) the panel will known which endpoints query to get such configuration. BTW this is exactly the current approach of DSpace <7 where each step have it own "processing class" (the backend that will stay on the REST side), the (JSP|XML)UI class (that will be moved on angular) and a set of properties or configuration files to use.

Finally, about the proposed endpoint URL restructure 
> /api/submission -> List of collection IDs in which you are authorized to submit

yes, we will need an endpoint to expose the collections where the submission can be done but this is not something that should go with the configuration staff, this is why I haven't included it in this PR. It could be a "filter" on the collections endpoint (aka /api/core/collections/search/authorized?action=ADD) or a dedicated endpoint. I tend to prefer the second option as it allows us eventually to evolve the concept of collections where you can submit away from the ACL ADD permission

> /api/submission/<coll-id> -> List of panels (steps) for a submission in this collection

The submission configuration is clearly a REST resource, right now stored in the item-submission.xml file in future we can decide for a different backend (db, spring file, etc.). This mean that our convention and the REST best practice suggest to have an endpoint to expose the collection of these resources (/api/configuration/submission-definitions) and provide access to a single configuration adding the /:id (that mean the submission name at the current stage). I see in your next comment that you have already found how I propose to implement such functionality with a search method


> /api/submission/<coll-id>/<panel-number> -> JSON representation of all the fields with their description present on the current panel.

as detailed below to access an element in a resource collection we need to use URL-of-the-collection/:id
right now we don't have an identifier for the single panel I manage that as an addressable resource only when we will find a need for that

Reply:
@abollini : I agree that the merger of input-forms and item-submission concepts doesn't need to happen in DSpace 7. However, it *is* worth considering whether keeping them as separate endpoints makes the logic in the Angular UI slightly more complex.

As two endpoints, it seems like the Angular UI will need to know to make a secondary REST request on "panels" that require an input-form configuration (i.e. `type=inputform`).  

So, it is worth considering whether some of this parsing of backend configurations could arguably be more easily achieved in Java code. As we've already decided, we want to push as much business logic down into the REST API as possible. 

To clarify here, I don't feel strongly either way...but I felt it is worth asking the question of whether we are accidentally making the code/logic in the Angular UI more complex by keeping these endpoints separate.

Reply:
it is not only a question about effort / timing. Each panel type will need a specific configuration that cannot be bound in advance in a specific structure (as new panel type must be pluggable in the system to allow extension in the same way that we currently support for Submission Step). It is easier in my opinion to left to the panel implementation the responsibility to retrieve the additional information (configuration) that it needs from whatever resource it want.